content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md

```diff
@@ -0,0 +1,99 @@
+---
+layout: blog
+title: 'Annotating Kubernetes Services for Humans'
+date: 2021-02-27
+slug: annotating-k8s-for-humans
+---
+
+**Authors:** Richard Li (Ambassador Labs)
+
+Have you ever been asked to troubleshoot a failing Kubernetes service and struggled to find basic information about the service such as the source repository and owner?
+
+One of the problems as Kubernetes applications grow is the proliferation of services. As the number of services grows, developers start to specialize working with specific services. When it comes to troubleshooting, however, developers need to be able to find the source, understand the service and dependencies, and chat with the owning team for any service.
+
+## Human Service Discovery
+
+Troubleshooting always begins with information gathering. While much attention has been paid to centralizing machine data (e.g., logs, metrics), much less attention has been given to the human aspect of service discovery. Who owns a particular service? What Slack channel does the team work on? Where is the source for the service? What issues are currently known and being tracked?
+
+## Kubernetes Annotations
+
+Kubernetes annotations are designed to solve exactly this problem. Oft-overlooked, Kubernetes annotations are designed to add metadata to Kubernetes objects. The Kubernetes documentation says annotations can “attach arbitrary non-identifying metadata to objects.” This means that annotations should be used for attaching metadata that is external to Kubernetes (i.e., metadata that Kubernetes won’t use to identify objects. As such, annotations can contain any type of data. This is a contrast to labels, which are designed for uses internal to Kubernetes. As such, label structure and values are [constrained](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) so they can be efficiently used by Kubernetes.
+
+
+## Kubernetes Annotations in Action
+
+Here is an example. Imagine you have a Kubernetes service for quoting, called the quote service. You can do the following:
+
+```
+$ kubectl annotate service quote a8r.io/owner=”@sally”
+```
+
+In this example, we’ve just added an annotation called `a8r.io/owner` with the value of @sally. Now, we can use `kubectl describe` to get the information.
+
+```
+$ kubectl describe svc quote
+Name:              quote
+Namespace:         default
+Labels:            <none>
+Annotations:       a8r.io/owner: @sally
+Selector:          app=quote
+Type:              ClusterIP
+IP:                10.109.142.131
+Port:              http  80/TCP
+TargetPort:        8080/TCP
+Endpoints:         <none>
+Session Affinity:  None
+Events:            <none>
+```
+
+If you’re practicing GitOps (and you should be!) you’ll want to code these values directly into your Kubernetes manifest, e.g.,
+
+```
```

Comment:
If you want, append `yaml` to the block opener (to make <tt>\`\`\`yaml</tt>) and the website will syntax-highlight the example manifest.